learn/shorts/functional-programming-python.qmd

```diff
@@ -0,0 +1,215 @@
+```
+
+In this example, the `map` function is used to square each number in the
+list of numbers. The lambda function is used to define the operation to
+be applied to each element of the list, in this case by squaring it. The
+`map` function returns an iterator, so we convert it to a list to see
+the result.
+
+#### **`reduce`**
+
+This is a function in the `functools` module in Python that takes a
+function and an iterable as arguments and returns a single value by
+applying the function cumulatively to the elements of the iterable.
+
+##### Syntax:
+
+``` python
+reduce(function, iterable)
+```
+
+The `function` takes two arguments and is applied cumulatively to the
+elements of `iterable` to reduce them to a single value.
+
+It is slightly different from the first two functions in that it isn't
+built into Python but is available in the `functools` module.
+
+For example:
+
+``` python
+from functools import reduce
+numbers = [1, 2, 3, 4, 5]
+sum_of_numbers = reduce(lambda x, y: x + y, numbers)
+print(sum_of_numbers)
+```
+
+In this final example, the `reduce` function is used to calculate the
+sum of the numbers in the list. The lambda function defines how the
+elements of the list should be combined, in this case by adding them
+together. As the `reduce` function returns a single value, we can print
+it directly.
+
+### What does the lambda keyword do in Python?
+
+The `lambda` keyword in Python is used to create what is called
+anonymous functions (i.e. not named, and therefore bound to the place
+where they are defined). In functional programming, the `lambda`
+function is often used to define simple functions that are used only
+once.
+
+##### Syntax:
+
+``` python
+lambda arguments: expression
+```
+
+A `lambda` function can take any number of arguments but must have only
+a single expression, which is evaluated and returned.
+
+In the `filter` example above, the lambda function
+(`lambda x: x % 2 == 0`) is used to define the condition for filtering
+the list of numbers. The argument `x` represents each element of the
+list, and the expression `x % 2 == 0` checks if the element is even. The
+lambda function used in the `map` example (`lambda x: x ** 2`) is used
+in much the same way, with a single argument and a simple expression.
+
+In the `reduce` example, the lambda function (`lambda x, y: x + y`)
+takes two arguments and returns their sum. The first argument represents
+the accumulated value, and the second argument represents the next
+element of the list. The lambda function is applied to each pair of
+elements in the list until a single value is returned.
+
+### Functional Programming-Adjacent Tools
+
+Python provides several tools that, while not purely functional,
+facilitate a transition to functional programming principles:
+
+-   **List Comprehensions & Generator Expressions** – Provide a concise
+    syntax for transforming data iterables, though they are not strictly
+    functional due to potential side effects.
```

Comment:
Maybe in this section you could compare list comprehensions/for loops with map, and give some reasons why the map has advantages over the other two.